### DISCUSSION TITLE
egs++ compiling issue: Undefined symbols for architecture arm64

### DISCUSSION THREAD

**Discussion by walleludvig** [nrc-cnrc/EGSnrc#996] [Questions]
Hi EGSnrc community,

I am trying to implement a modified version of `egs_application` with egs++, but get a linking error at compilation. The issue arises when attempting to compile the original `egs_application.cpp` too (source code as distributed with EGSnrc install [https://nrc-cnrc.github.io/EGSnrc/doc/pirs898/egs__application_8cpp.html](url)).

The error is:
```
Undefined symbols for architecture arm64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture arm64
```

This error does not arise at compilation of any other applications (e.g., `egs_cavity`, `egs_app`, etc.).

Has anyone else experienced a similar issue? Anyone got some suggestions as to how I can resolve this?

I have gcc-13 compilers installed through home-brew and running on MacOS Ventura 13.4 (M1 chip).

**Comment by ftessier** [comment]
Well you cannot compile `egs_applicaton.cpp` directly, this is just a c++ library to build an application (so there is no `__main` symbol in the object code). If you want to modify `egs_application.cpp` and compile an application that uses your modified `EGS_Application` class. In that case, go into `$HEN_HOUSE/egs++` and issue the command `make`. This will rebuild the library object and place it in `egs++/dso/.../` so that it is in your path.

Better yet, keep the original `egs_application.cpp` and create your own copy of it under different `.cpp` and `.h` names under `egs++/`. Adjust the `Makefile` accordingly so that your new library object is included when you issue `make`.

> Note: to compile a file as a library object (without requiring a `__main` symbol), you need to use the gcc flag `-c`. But in this case, you would then have to move it to the right location etc. Better use `make`!

**Comment by walleludvig** [comment]
Hi @ftessier, thank you for the response. 

Compiling in egs++ did indeed work fine, however I am unsure of how I can manage to include the modified version of `egs_application`. When I issues `make` after moving the modified `egs_application` and its header file over to egs++ and modifying the `Makefile` I keep getting some errors like `*** make error: no rules to make target egs_my_app.h needed **`  when I add the header file to the same lists that I found `egs_application` in. However, I also attempted to exclude it in some of these and only mention it in the `egspp_files` and a line to generate the DSO1 object. I still then get an error about no` rules to make target ... needed by egs_my_app.o`. I believe an object file (`egs_my_app.o`) was created in the `egs++/dso/.../` directory regardless and I thought this file is what I am after all looking for? 

What I don't understand is how I can generate information/"rules" on how to make the target for the modified version of `egs_application`. Apologies for my lack of knowledge, but your help and expertise is greatly appreciated. Thank you!

**Reply by ftessier** [reply]
Yes, makefiles can be convoluted! To add `egs_my_app.cpp`, you have to:


### 1. Add `egs_my_app` in the `egspp_files` variable:

https://github.com/nrc-cnrc/EGSnrc/blob/5c981341fd3bb1f6b57d818e1ae4402ae3b23957/HEN_HOUSE/egs%2B%2B/Makefile#L40-L46

### 2. Create a rule to make `egs_my_app.o`, similar to what is done for `egs_application`:

https://github.com/nrc-cnrc/EGSnrc/blob/5c981341fd3bb1f6b57d818e1ae4402ae3b23957/HEN_HOUSE/egs%2B%2B/Makefile#L142

### 3. Add `egs_my_app.h` where `egs_application.h` is included in the Makefile:

https://github.com/nrc-cnrc/EGSnrc/blob/5c981341fd3bb1f6b57d818e1ae4402ae3b23957/HEN_HOUSE/egs%2B%2B/Makefile#L85-L86

https://github.com/nrc-cnrc/EGSnrc/blob/5c981341fd3bb1f6b57d818e1ae4402ae3b23957/HEN_HOUSE/egs%2B%2B/Makefile#L146-L148

https://github.com/nrc-cnrc/EGSnrc/blob/5c981341fd3bb1f6b57d818e1ae4402ae3b23957/HEN_HOUSE/egs%2B%2B/Makefile#L150-L152

### 4. Modify the class name

**However,** for all of this to work, you must rename the class in your new files (to `EGS_Myapp`, for example), otherwise the `EGS_Application` symbol will be _multiply defined._ So perhaps it is easier to just modify `egs_application.*` in the end? 🤔

**Comment by rtownson** [comment]
What are you actually trying to do? Are you sure that you need to modify egs_application, and it can't be done in your new application instead?

**Reply by ftessier** [reply]
Good point! Normally one would indeed _derive_ the `EGS_Application` class and add features there.

**Reply by walleludvig** [reply]
I am trying to compile an egs++ application code written by Jonas Lippuner in 2011 called Epp. He provided the source code and a user manual at this link: [https://jonaslippuner.com/research/epp/](url).

While the instructions for installation indicates that all that is needed is to cd into the installed directory and do `make`. This failed for me and so I am trying to make it work step by step instead. I also suspect that since the user code is from 2011 it may be a bit outdated and need modifications along the way.

**Reply by rtownson** [reply]
I'd just like to point out that we've discussed solving the epp compatibility with modern EGSnrc before: https://github.com/nrc-cnrc/EGSnrc/discussions/746

**Reply by walleludvig** [reply]
Interesting! Thanks for pointing that out @rtownson. I guess I am currently making my own attempt at making it work with modern EGSnrc at the moment and while it seems like I have got it compiled I am still struggling a bit to make it run a simulation. As I am only learning about C++ on the fly here I've got doubts in my implementation of it and assume that the errors I am getting got to do with the way I managed to get it to "compile".

I will let you know if I get any success.

**Comment by walleludvig** [comment]
Thank you so much for the descriptive instructions for help @ftessier. Really appreciate your assistance!

I ended up modifying egs_application directly instead (but with clear comments to identify where about). These modifications seem to have compiled fine.

However when I try to compile the actual egs++ application which uses this modified version of egs_application, I am still getting some linking errors with a list of `duplicate symbol...` (total of 47 duplicate symbols between `Epp_macosx.o` and `egs_advanced_application_macosx.o`) It is also followed by the following error `ld: 47 duplicate symbols for architecture arm64
collect2: error: ld returned 1 exit status`.

Any chance you happen to know why this is happening as well?

**Reply by walleludvig** [reply]
I think I've actually managed to fix this issue by changing one of the `#include egs_advanced_application.cpp` to `#include egs_advanced_application.h`. `Epp` seems to have compiled successfully now but when running one of the examples I get an error saying:
```
You must turn ON Compton binding corrections when using
a detailed atomic relaxation (eadl_relax=true) since 
binding energies taken from incoh.data below 1 keV!
```

Which I found to be a result of `eadl_relax & photon_xsections = 'xcom'` returning `True` in `egsnrc.mortran`. While I doubt this is an instance of something like incorrect compilation I cannot be sure as I do not understand enough of all this.
Furthermore I do not really understand how to resolve the issue. I've tried changing multiple of the MC Transport parameters such as `Photon cross sections` to others than `xcom` but I just end up with a` segmentation fault`.

Apologies for constantly asking, but any suggestions on how to fix this Compton binding corrections issue? How can I do as the error message suggest and turn ON Compton binding corrections?

**Reply by rtownson** [reply]
In the input file, you could remove the transport parameters to just use the defaults, or set `Bound Compton scattering=On`. At least this is the solution for other egs++ applications..

**Reply by walleludvig** [reply]
@rtownson 

Setting `Bound Compton Scattering=On` does not change results and I receive the same error.

Setting all the MC Transport parameters to default values (Jonas' code requires that a block of MC transport parameters is included in input file) results in a segmentation fault.

I will keep trying to make this work, but any suggestions are greatly appreciated!

**Comment by walleludvig** [comment]
Hi @rtownson and @ftessier,

Just writing here to ask if either of you may know why the following issue I am having might be occurring:

While Epp seems to have compiled fine for me know, when I try to run a simulation with it (`Example_1.egsinp` or `Example_2.egsinp` provided by Jonas) I keep getting a `segmentation fault` error. 

I have noticed that when running a simulation a new input file (with filename eg `Example_1.egsinp.egsinp`) is created in my directory and this file just contains the `MC transport parameters`. I therefore suspect that I am getting a `segmentation fault` because the simulation starts reading from this file at runtime instead. However I am unsure as the creation of geometries, sources, media etc seems to have been successful during compilation of the input file (i.e., the output to terminal is as expected and the error only occurs when trying to execute the first batch).

Have you encountered or seen any similar behaviour before?

Thanks for all your help so far!

**Reply by mxxo** [reply]
Hi @walleludvig, 

I'm not sure about your specific issue but I have found building EGSnrc with Address Sanitizer to be very helpful for debugging segfaults. 

**To build EGSnrc with Address Sanitizer support**

(Assuming your EGSnrc configuration is called `linux`)

* In `HEN_HOUSE/specs/linux.conf` change `FOPT` and `C_FLAGS` to:
```
FOPT = -O2 -mtune=native -g -fsanitize=address -static-libasan
...
C_FLAGS = -O2 -fPIC -g -fsanitize=address -static-libasan
```
* In `HEN_HOUSE/specs/egspp_linux.conf` change `opt` to:
```
opt = -O2 -mtune=native -g -fsanitize=address -static-libasan
```

Then recompile `egs++` and your application: 

```
cd HEN_HOUSE/egs++ && make realclean && make
cd HEN_HOUSE/user_codes/your_application && make clean && make # or wherever your application is
```

Once a segfault occurs there should be a log and backtrace printed pointing to the code that is causing the segfault.

**Reply by walleludvig** [reply]
Hi @mxxo,

Thank you very much for your great suggestion!

I have tried to use Address Sanitizer as suggested and in a couple different ways, however I using MacOS (Ventura) with Xcode command line tools (14.3) and I get a linker error that I have not been able to resolve.

For reference the `-static-libasan` is outdated and not required anymore [https://stackoverflow.com/questions/37970758/undefined-reference-errors-when-trying-to-use-address-sanitizer-with-gcc/40215639#40215639](url).

Therefore by following your suggestions as to which configs and flags I added

```
FOPT = -O2 -mtune=native -g -fsanitize=address
C_FLAGS = -O2 -fPIC -g -fsanitize=address
```
and 
`opt = -O2 -mtune=native -g -fsanitize=address`

but at compilation of `egs++` everything seems fine up until the last moment when the following linker error occurs 
```
ld: library not found for -lasan
collect2: error: ld returned 1 exit status
make: *** [/Users/ludvigwalle/EGSnrc/HEN_HOUSE/egs++/dso/osx/libegspp.dylib] Error 1
```
From reading various forums online I believe the problem lies at compatibility between Xcode and Homebrew's gcc but with minimal knowledge on these topics I have not been able to resolve it.

If you do have any suggestions I would greatly appreciate advice as this debugger would likely be extremely useful for my situation.

Regardless, thank you very much for the suggestion!

**Reply by mchamberland** [reply]
@walleludvig @mxxo  For what it's worth, I also have not been able to get address sanitizer to work on macOS, using the Homebrew gcc.

**Reply by walleludvig** [reply]
@mchamberland Thanks for letting us know. Knowing that I am probably not gonna spend anymore time trying (was tempted to try installing EGSnrc with Xcode's gcc compilers but with great hesitancy).

On another note, any chance you've tried using the `eg++` user code `Epp` on your mac? As mentioned above I believe I have managed to get it to compile but not gotten it to actually run simulations yet. Happy to share my files if you'd wanna attempt it.

**Reply by mchamberland** [reply]
@walleludvig I've taken a look at the epp code. Like @ftessier mentioned, it's good code, but a little of a mess because of how it handles modifications to egs_application and things like that. I have half a mind to try cleaning it up based on my experience with egs_brachy. But that's not trivial work.

**Reply by walleludvig** [reply]
@mchamberland

Yes okay I see. As also suggested by @ftessier I ended up modifying `egs_application` itself to implement the stuff needed by `Epp` and then instead of including `egs_advanced_application.cpp` in `Epp.h` I changed it to include only `egs_advanced_application.h` and so it seems to have compiled. Presumably there is still some issues I have not addressed as it still results in `segmentation fault` when running the simulation.

I may have to find another route to achieve simulating an x-ray radiography setup for my case.

Thanks very much for having a look. I appreciate your responses and efforts🙏

**Reply by ftessier** [reply]
`Epp` could also be reworked and included as a submodule in the new `third-party` directory, if someone wants to take owner ship. As mentioned in #746, I think the original author would be supportive of that, or else including it in EGSnrc proper. Note that this is an item in the list of projects in #790.

**Reply by mchamberland** [reply]
@ftessier I might take a stab at it over the next couple of weeks. @walleludvig, I'll keep you posted.

**Reply by mchamberland** [reply]
@walleludvig @ftessier Good news! I updated Epp and got it running. I can run the first example.

I stripped out a lot of stuff that I thought was unnecessary, such as pre-processing input files for egsphant support. Egsphants are supported out of the box in EGSnrc and the CLRP fork of EGSnrc has an even better way to use egsphants (no ramp file needed!), so I think this is for the better.

At the moment, I've also taken out all the help messages (from using -h on the command line). I didn't feel like dealing with this tonight.

@walleludvig  Any chance you can send me an input file to see if it crashes on my end?

**Reply by mchamberland** [reply]
@walleludvig @ftessier  I was also able to run the second Epp example, with egsphant support provided by the CLRP version of egs_glib. So, everything appears to be working on my end. I don't have anything to benchmark it against, though.

**Reply by walleludvig** [reply]
Hi @mchamberland,

Apologies that I have not followed the progress you've made with updates on `Epp` application. I have not been able to log on for a while due some personal circumstances. However, it is with great pleasure to hear that you have successfully implemented both of the example input files provided with `Epp`! Thank you so much for taking the time to update this application; it will be a very valuable tool for my simulation needs and I would love to test it out.

I am not sure if my input files will be any better for verification testing than those already provided by Jonas with the `Epp` package, however I have attached a zip directory with the simulation setup I was planning to do my first experimental tests with. I wrote this for implementation in `egs_chamber` and it has worked for me, but as I have not been able to trial and error with the `Epp` application I am not sure if the setup would work straight out of the box, apologies. The input file I have included is simply the one I used for `egs_chamber`. Please let me know if I should make some changes to the input file.
[Epp_trial_inputs.zip](https://github.com/nrc-cnrc/EGSnrc/files/12352645/Epp_trial_inputs.zip)

Thank you so much for making the time to work on this application. I would really love to test it out if you would like to share the files or explain what you've modified in the originals to make it work with modern EGSnrc.

**Reply by mchamberland** [reply]
@walleludvig My version of `Epp` is [available here](https://github.com/mchamberland/Epp). You would want to place those files within `egs_home/Epp`, then typing `make` within that folder should build the application.

Note that the documentation is not yet up to date. I removed the custom egsphant support, since EGSnrc supports egsphant out of the box now.

I'll take a look at your input files.

**Reply by mchamberland** [reply]
@walleludvig After adding the relevant media definitions and Epp-specific output options, the input file runs fine. So, feel free to give my version of Epp a shot.

**Reply by walleludvig** [reply]
Amazing! Thank you so much for providing this @mchamberland. I will test out your new version of Epp and I am sure it will a very helpful tool for my needs!

**Reply by walleludvig** [reply]
Hi @mchamberland,

I have successfully trialed your new version of Epp for the example input files and it is working very well with output detector images produced as expected. Thank you again for doing this!

However when I try with my own simulation setup I cannot seem to get the expected output images produced and I am only getting empty images as output. I believe there may be some issues with how I am setting up the sources and that ultimately the photons aren't reaching the detector. For this purpose I have been trying to use `ptracks` in `ausgab object definition` but it seems that Epp does not support this. I have tried using it for `Example_1.egsinp` but no tracks are stored in the output `Example_1.ptracks` file.

Do you happen to know if use of `ptracks` is compatible with this version of Epp, or if there may be some simple solution to making it work?

**Reply by mchamberland** [reply]
@walleludvig Tracks should be supported out of the box. I was able to generate some right now (though I have not visualized them).

Maybe double-check your input file?

**Reply by walleludvig** [reply]
@mchamberland Okay, would have to be something I am getting wrong then. However I did use the exact same code in `egs_chamber` and got the tracks are visualised just fine. I will look more into the output `.ptracks` file to see if the issue is with visualisation in `egs_view` or writing to the `.ptracks` file.

Thanks for checking.

**Comment by morfius84** [comment]
I have a problem when compiling dosrznrc input

```
Error: Symbol 'git_hash' at (1) has no IMPLICIT type
mingw32-make[1]: *** [D:\EGSnrc\EGSnrc-master\HEN_HOUSE\makefiles\standard_makefile:90: D:\EGSnrc\EGSnrc-master\third-party\bin\win3264\dosrznrc.exe] Error 1
mingw32-make: *** [D:\EGSnrc\EGSnrc-master\HEN_HOUSE\makefiles\standard_makefile:189: opt] Error 2
```
